src/smc-webapp/r_misc/selector-input.tsx

```diff
@@ -0,0 +1,82 @@
+import * as React from "react";
+import * as misc from "smc-util/misc";
+
+const { FormControl, FormGroup } = require("react-bootstrap");
+
+interface Props {
+  options:
+    | string[]
+    | { value: string; display: JSX.Element }[]
+    | { [keys: string]: JSX.Element };
+  disabled?: boolean;
+  selected?: string;
+  on_change?: (selected: string) => void;
+}
+
+// If the first element is a string, we assume the rest to be a string
+function isStringArrayHeuristic(a: any): a is string[] {
```

Comment:
What does `a is string[]` mean for the return type?  Shouldn't the return type be `boolean`???

I'm not doubting this is valid typescript; it's just that I don't understand it....

Reply:
This is the syntax to tell Typescript that is function is a typechecker. eg
```
// foo: string[] | number;
let length: number = 0;
if (isStringArrayHeuristic(foo)) { // Behaves as if we could do typeof foo == string[]
  length: number = foo.length;  // <-- TS knows foo is a string array so has property length 
} else {
  length = foo; // <-- TS knows foo is a number
}
```

Reply:
(and this code is all merged into master and live on prod already, so this is purely for my education.)

Reply:
In the documentation is called a "User-Defined Type Guards".

It's on this page https://www.typescriptlang.org/docs/handbook/advanced-types.html